### ISSUE TITLE
AUR qtile-git package fails when building following wlroots 0.16 update

### ISSUE THREAD

**Issue by Qpernicus** [qtile/qtile#4108]
### The issue:

I used both Paru and Yay to update qtile-git, but both fail:

The Version is qtile-git-0.20.0.r33.gc1f50739-2

The Errors vary a little:

```
build/temp.linux-x86_64-cpython-310/libqtile.backend.wayland._libinput.c:571:10: fatale fout: wlr/backend.h: File or folder doesn't exist.
  571 | #include <wlr/backend.h>
      |          ^~~~~~~~~~~~~~~
compilatie is beëindigd.
error: command '/usr/bin/gcc' failed with exit code 1
```

```
build/temp.linux-x86_64-cpython-310/libqtile.backend.wayland._libinput.c:603:10: fatale fout: wlr/types/wlr_surface.h: File or Folder doesnt exist
  603 | #include <wlr/types/wlr_surface.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~
compilation ended.
error: command '/usr/bin/gcc' failed with exit code 1
```

Version pywlroots: python-pywlroots 0.15.24-3
Version Wlroots: wlroots0.15 0.15.1-1 

But when i try and use YAY it cant find the update....

### Required:

- [X] I have searched past issues to see if this bug has already been reported.

**Comment by Qpernicus** [comment]
I just realised something.. when i do yay -Syu... qtile-git doesn't popup... 


but this does:

:: Searching AUR for updates...
 -> Flagged Out Of Date AUR Packages:  paru 

Could that cause the error in building Qtile?

**Comment by Sydiepus** [comment]
Both errors suggest that wlroots is missing. 
Can you check if there is a folder named `wlr` inside `/usr/include` if no then wlroots is probably not installed, else you should be able to find the files. 

Not sure about using different AUR helpers but maybe they are causing the problem. 
Try deleting qtile from one and install it from the other.

**Comment by Qpernicus** [comment]
> Both errors suggest that wlroots is missing. Can you check if there is a folder named `wlr` inside `/usr/include` if no then wlroots is probably not installed, else you should be able to find the files.
> 
> Not sure about using different AUR helpers but maybe they are causing the problem. Try deleting qtile from one and install it from the other.

The folder isn't there although wlroots0.15 is installed (reinstalled via pacman, reinstalled via yay, no difference) Both packagemanagers warn they about to REINSTALL the package, eventhough the above folder isn't there.

**Comment by Qpernicus** [comment]
when searching with yay: yay -Ss qtile-git

aur/qtile-git 0.20.0.r33.gc1f50739-2 (+70 0.03) (Installed: 0.22.1.r101.g6a37adca-1)

**Comment by elParaguayo** [comment]
Ignore the version reported by the AUR. That's just the version when the PKGBUILD was uploaded. It will always install the latest git version.

**Comment by Sydiepus** [comment]
As elParaguayo said you can ignore qtile's version that is reported by the AUR.
Qtile isn't the cause of the problem but your system not installing wlroots correctly.
Try uninstalling wlroots from pacman and any other aur helper and install it using one of them making sure the folder appear. Try to test each one separately to find the one that installs it.

**Comment by Qpernicus** [comment]
Ok, i am a idiot.. but it might have shown me the error as well...

I went to /usr/include and did an LS.. No WLR visible............but i looked at the FILES... IN FOLDERS. which are higher up in the list.... there is the following:

…/include/wlroots0.15/wlr 

So could it be that the build fails because it is looking at include/wlr instead of wlroots0.15/wlr/ >> Because the latter contains :

drwxr-xr-x    - root  8 jan 08:26 backend
drwxr-xr-x    - root  8 jan 08:26 interfaces
drwxr-xr-x    - root  8 jan 08:26 render
drwxr-xr-x    - root  8 jan 08:26 types
drwxr-xr-x    - root  8 jan 08:26 util
.rw-r--r-- 2,1k root  1 jan 19:41 backend.h
.rw-r--r--  243 root  1 jan 19:41 config.h
.rw-r--r--  272 root  1 jan 19:41 version.h
.rw-r--r-- 3,9k root  1 jan 19:41 xcursor.h
.rw-r--r-- 9,0k root  1 jan 19:41 xwayland.h


So the files/folders it needs.. I don't think its a problem root owns them??

**Comment by Jareng** [comment]
The wayland dependencies have been removed from qtile-git you should use qtile-wayland-git

**Comment by Qpernicus** [comment]
> The wayland dependencies have been removed from qtile-git you should use qtile-wayland-git

Yeah, well, thats what i thought, but that won't work, because that won't build either, because of the same error like above

**Comment by Jareng** [comment]
I can't build it either, But i don't have the same error as you
arch upgraded wlroots to 0.16 a few days ago and renamed the 0.15 version to wlroots0.15, and it looks like pywlroots need 0.15

**Comment by Qpernicus** [comment]
> I can't build it either, But i don't have the same error as you arch upgraded wlroots to 0.16 a few days ago and renamed the 0.15 version to wlroots0.15, and it looks like pywlroots need 0.15

Yeah, and i have 0.15 installed, but i have the feeling that because the renaming the build is looking in usr/include/wlr instead of usr/include/wlroots0.15/wlr .. which is why it cant find it. But thats an assumption.. 

I changed qtile-git to qtile, that worked, but now i cant go back to qtile-git or qtile-wayland-git so i might have f#@ked myself..

**Comment by elParaguayo** [comment]
Sounds like we may need @m-col to update the AUR package as the dependencies have changed.

You can always download the PKGBUILD and edit the dependencies yourself (eg use the wlroots0.15 package).

Pywlroots should hopefully support 0.16 soon.

**Comment by Qpernicus** [comment]
> Sounds like we may need @m-col to update the AUR package as the dependencies have changed.
> 
> You can always download the PKGBUILD and edit the dependencies yourself (eg use the wlroots0.15 package).
> 
> Pywlroots should hopefully support 0.16 soon.

But .. when i look here:

https://aur.archlinux.org/packages/qtile-wayland-git?all_deps=1#pkgdeps

Wlroots is not a dependency, only pywlroots.. so, that means the problem is in pywlroots really.. 

But there (https://archlinux.org/packages/community/x86_64/python-pywlroots/)  the dependency is wlroots0.15...

**Comment by Qpernicus** [comment]
I'm getting a bit chicken and egg feeling here.. especially since i removed the qtile-git and now can't build it (nor the wayland version).. Ill spend my time on Hyprland for a while then..

**Comment by elParaguayo** [comment]
I suspect m-col can sort this pretty quickly as he's the main developer for our Wayland backend. Sorry, I'm just not as familiar with it.

**Comment by m-col** [comment]
I see the official repos `python-pywlroots` package is using CFLAGS and LDFLAGS to inform the build of the `/usr/include/wlroots0.15/` folder where the headers are (https://github.com/archlinux/svntogit-community/commit/bcdd93349bb9565e730019576439898cbfb259c6). That change was presumably sufficient to get pywlroots to build fine, so perhaps it's enough to get the `qtile-git` and `qtile-wayland-git` packages to build fine too. If you make the same changes shown in that diff to the qtile-git/qtile-wayland-git PKGBUILD before building, does that fix it?

**Comment by Qpernicus** [comment]
> 

To make sure: I did the following,

sudo nvim PKGBUILD fo the file in folder .cache/yay/qtile-wayland-git 

i scrolled to the line: 

build() {

and pasted these at the top of the 'block' :

 ```
 export CFLAGS="$CFLAGS -I/usr/include/wlroots0.15"
  export LDFLAGS="$LDFLAGS -L/usr/lib/wlroots0.15"
```
  
  this I changed as well:
  
pkgrel=1
pkgrel=2

These i judged as not relevant? Am i correct?

```
depends=(glibc python-cffi python-pywayland python-xkbcommon wlroots
libwlroots.so)

````depends=(glibc python-cffi python-pywayland python-xkbcommon libwlroots.so=10)`
  
  
  (I never editted a PKGBUILD file before..sorry)

I still get this error:

build/temp.linux-x86_64-cpython-310/libqtile.backend.wayland._libinput.c:571:10: fatale fout: wlr/backend.h: Bestand of map bestaat niet
  571 | #include <wlr/backend.h>
      |          ^~~~~~~~~~~~~~~
compilatie is beëindigd.
error: command '/usr/bin/gcc' failed with exit code 1

**Comment by Sydiepus** [comment]
Not sure if this will help but i tried to install `qtile-wayland-git` and the problem was that wlroots was not found during the build of libinput. 
I symlinked `/usr/include/wlr` to `/usr/include/wlroots0.15/wlr` and `/usr/lib/wlroots` to `/usr/lib/wlroots0.15`

I never even had problems with the headers not being found which makes things confusing.

**Comment by Jareng** [comment]
I tried the changes in the PKGBUILD, but I got the same errors as before

a ton of errors like that, all about `libqtile.backend.wayland._libinput.c`
```
/usr/include/wlr/types/wlr_surface.h:1:2: warning: #warning "wlr/types/wlr_surface.h has been deprecated and will be removed in the future. Use wlr/types/wlr_compositor.h and wlr/types/wlr_subcompositor.h." [-Wcpp]
    1 | #warning "wlr/types/wlr_surface.h has been deprecated and will be removed in the future. Use wlr/types/wlr_compositor.h and wlr/types/wlr_subcompositor.h."
      |  ^~~~~~~
build/temp.linux-x86_64-cpython-310/libqtile.backend.wayland._libinput.c: In function ‘_cffi_checkfld_struct_wlr_compositor’:
build/temp.linux-x86_64-cpython-310/libqtile.backend.wayland._libinput.c:2766:39: error: ‘struct wlr_compositor’ has no member named ‘subcompositor’
 2766 |   { struct wlr_subcompositor *tmp = &p->subcompositor; (void)tmp; }
      |                                       ^~
build/temp.linux-x86_64-cpython-310/libqtile.backend.wayland._libinput.c: In function ‘_cffi_checkfld_struct_wlr_drag_icon’:
build/temp.linux-x86_64-cpython-310/libqtile.backend.wayland._libinput.c:2859:33: error: ‘struct wlr_drag_icon’ has no member named ‘surface_destroy’
 2859 |   { struct wl_listener *tmp = &p->surface_destroy; (void)tmp; }
```

i have both wlroots 0.16 and wlroots0.15 installed

**Comment by Jareng** [comment]
The modified PKGBUILD fixed the issue for me

**Comment by Qpernicus** [comment]
> The modified PKGBUILD fixed the issue for me

What exactly did you do then?

**Comment by Jareng** [comment]
git clone https://aur.archlinux.org/qtile-wayland-git
cd qtile-wayland-git
add the two export under build in the PKGBUILD
makepkg -si

**Comment by Qpernicus** [comment]
> git clone https://aur.archlinux.org/qtile-wayland-git cd qtile-wayland-git add the two export under build in the PKGBUILD makepkg -si

Thx, that seemed to have done the trick!

**Comment by Qpernicus** [comment]
OK, so everything seems fine, only YAY and PARU don't agree on what needs to be updated.. but thats unrelated to this and might be caused by paru being marked as out of date.

in the future i'll see if a new update of either qtile or (py)wlroots works.. but thats no reason to keep this open

**Comment by m-col** [comment]
Thanks for confirming that the patch works. I'll update the AUR pkgbuilds.

**Comment by Qpernicus** [comment]
> Thanks for confirming that the patch works. I'll update the AUR pkgbuilds.

Not sure if this is related and how i could check.. but since the update; everytime i log in on fresh boot (SDDM) my qtile 'breaks' (Black screen, non-blinking cursor on screen) and i have to CTRL-ALT-DEL/Reboot and then it works fine.. 

So its a fresh boot, thats the only difference with second boot. But how could i  troubleshoot this?

**Comment by m-col** [comment]
Just to confirm, this is after updating the package again following the change I made to the pkgbuild to add those export lines? And also that this is with the latest package for python-pywlroots and by extension the wlroots0.15 package installed (and NOT the wlroots package)?

**Comment by Qpernicus** [comment]
> Just to confirm, this is after updating the package again following the change I made to the pkgbuild to add those export lines? And also that this is with the latest package for python-pywlroots and by extension the wlroots0.15 package installed (and NOT the wlroots package)?

- It is by manually editting the PKGBUILD (off the downloaded qtile-wayland-git (Installed: 0.22.1.r102.gf0d07ecf-1)) So no reinstalling of the package.. Could try that?
- Python-pywlroots  0.15.24-3 (No later version available?)
- the wlroots0.15 package is the only one installed

**Comment by m-col** [comment]
> It is by manually editting the PKGBUILD (off the downloaded qtile-wayland-git (Installed: 0.22.1.r102.gf0d07ecf-1)) So no reinstalling of the package.. Could try that?

I pushed an update to the AUR package, so if you update the package it should fetch that and install fine without your manual changes. This shouldn't change anything though, my update only added the same export lines that you added yourself.

> * Python-pywlroots  0.15.24-3 (No later version available?)
> 
> * the wlroots0.15 package is the only one installed

All good, thanks.

> So its a fresh boot, thats the only difference with second boot. But how could i troubleshoot this?

Strange that it works after a reboot! Display managers often have issues with wlroots. If you try running qtile without the display manager from the tty directly, does it work fine?

**Comment by Qpernicus** [comment]
> > It is by manually editting the PKGBUILD (off the downloaded qtile-wayland-git (Installed: 0.22.1.r102.gf0d07ecf-1)) So no reinstalling of the package.. Could try that?
> 
> I pushed an update to the AUR package, so if you update the package it should fetch that and install fine without your manual changes. This shouldn't change anything though, my update only added the same export lines that you added yourself.
> 
> > * Python-pywlroots  0.15.24-3 (No later version available?)
> > * the wlroots0.15 package is the only one installed
> 
> All good, thanks.
> 
> > So its a fresh boot, thats the only difference with second boot. But how could i troubleshoot this?
> 
> Strange that it works after a reboot! Display managers often have issues with wlroots. If you try running qtile without the display manager from the tty directly, does it work fine?

Ok, only the first try, but i started my laptop (SDDM-git) went to tty (F2) en started (qtile start -b wayland) .
Up until now (15mins) its still holding.. So indeed, maybe SDDM suddenly didn't like the change to qtile-wayland-git (or changes in wlroots/pywlroots..

**Comment by Qpernicus** [comment]
Update after 3 boots (4 to be fair)

- 2 boots to tty (like above) --> No Problem
- 1 boot via SDDM-git to Qtile (Wayland) --> From SDDM to a 'live' Qtile takes a long time with a black screen and blinking cursor
- 1 boot via tty --> No Problem..

So SDDM doesnt seem to play nice after i migrated from qtile-git to qtile-wayland-git.. Not sure where that problem lies though.. In this issue, of in SDDM?

**Comment by Qpernicus** [comment]
Ok.. Weirdly.. it has been fine now the last 2 starts.. no updates that i know of..
